### PR TITLE
Use Legion with CMake's native CUDA language

### PR DESCRIPTION
**Pull request by trxcllnt**
Redo of https://github.com/nv-legate/legate.core/pull/725 that touches fewer CI-related files. Updates Legion to the commit hash in https://github.com/nv-legate/legate.core/pull/827.

### CODE REVIEW

File: install.py [hunk 1]
```diff
@@ -594,7 +674,7 @@ def driver():
         dest="arch",
         action="store",
         required=False,
-        default="NATIVE",
+        default="all-major",
```
Comment (manopapad) [inline install.py:677]: It looks like this argument no longer accepts codenames, e.g. "volta"?

**Comment by Jacobfaib** [comment]
@trxcllnt If not already, can you also add a CI job(s) that tests the editable install + venv in this PR?

File: install.py (hunk 1 above)
Comment (trxcllnt) [reply install.py:677]: Yeah, CMake (and nvcc) accept `native` and not `NATIVE`. The uppercase version was a rapids-cmake thing, but now we have support directly in CMake and nvcc.

File: install.py [hunk 2]
```diff
@@ -68,9 +68,11 @@ def __call__(self, parser, namespace, values, option_string):
         setattr(namespace, self.dest, not option_string.startswith("--no"))
 
 
-def execute_command(args, verbose, **kwargs):
+def execute_command(args, verbose, ignore_errors=False, **kwargs):
     if verbose:
         print(f"Executing: {' '.join(args)} with {kwargs}")
+    if ignore_errors:
+        subprocess.call(args, **kwargs)
     subprocess.check_call(args, **kwargs)
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply install.py:76]: Oh good catch. I think there was a weird merge here.

File: scripts/generate-conda-envs.py [hunk 3]
```diff
@@ -71,18 +71,32 @@ def conda(self) -> Reqs:
...
+
+        if self.compilers:
+            if V(self.ctk_version) < V("12.0"):
+                deps += (f"nvcc_linux-64={self.ctk_version}",)
             else:
-                deps += (
-                    "gcc_linux-64=11.*",
-                    "gxx_linux-64=11.*",
-                )
+                deps += ("cuda-nvcc",)
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply scripts/generate-conda-envs.py:85]: nvcc isn't available on windows?

[the remark this answers is not in the document]
Comment (manopapad) [reply scripts/generate-conda-envs.py:85]: You're right, the "else" branch is OK. But the "if" branch mentions the package `nvcc_linux-64`, which is (presumably) linux-specific.

File: scripts/generate-conda-envs.py [hunk 4]
```diff
@@ -145,7 +159,7 @@ def conda(self) -> Reqs:
         return (
             "cffi",
             "llvm-openmp",
-            "numpy>=1.22",
+            "numpy>=1.22,<1.25",
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply scripts/generate-conda-envs.py:162]: IIRC at one point a few months ago numpy 1.25 caused a weird error. I'll lift the restriction and see if it works.

File: install.py (hunk 1 above)
Comment (manopapad) [reply install.py:677]: OK, I'll switch to using "native". But what about the codenames ("volta", "turing" etc.)? Should we switch to using architecture numbers instead?

File: scripts/generate-conda-envs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (trxcllnt) [reply scripts/generate-conda-envs.py:85]: Oh right heh

[the remark this answers is not in the document]
Comment (trxcllnt) [reply scripts/generate-conda-envs.py:85]: Do we even support building on Windows?

File: install.py [hunk 5]
```diff
@@ -477,8 +533,21 @@ def validate_path(path):
         cmake_flags += [f"-Dlegate_core_LEGION_BRANCH={legion_branch}"]
     if build_docs:
         cmake_flags += ["-Dlegate_core_BUILD_DOCS=ON"]
+    if build_tests:
+        cmake_flags += ["-Dlegate_core_BUILD_TESTS=ON"]
     if build_examples:
-        cmake_flags += ["-Dlegate_core_EXAMPLE_BUILD_TESTS=ON"]
+        cmake_flags += ["-Dlegate_core_BUILD_EXAMPLES=ON"]
+
+    if not is_conda and editable:
```
[the remark this answers is not in the document]
Comment (trxcllnt) [reply install.py:541]: I just did what seemed to work in a conda env and pip venv, but I agree the logic is super weird.

Note: `CMAKE_INSTALL_PREFIX=<skbuild_dir>/cmake-install` for all the commands below.

### In a pip virtual env

We install `_deps/legion-build/bindings/python` into `${CMAKE_INSTALL_PREFIX}` (i.e. `<skbuild_dir>/cmake-install`).

This produces the following cmake install rule for Legion's python bindings:
```shell
${VIRTUAL_ENV}/bin/python3 \
    -m pip install \
    --global-option=--cmake-build-dir "--global-option=<skbuild_dir>/cmake-build/_deps/legion-build/runtime" \
    --global-option=--prefix "--global-option=${CMAKE_INSTALL_PREFIX}" \
    --ignore-installed \
    --root / \
    --prefix ${CMAKE_INSTALL_PREFIX} .
```

Because there's no `--editable` flag, pip installs the Legion Python bindings into `${VIRTUAL_ENV}/<..>/site-packages`.

During this process, the Legion Python bindings' `setup.py` inlines the `--prefix` arg into the generated `legion_canonical_cffi.py`:
```python
libpath = "<skbuild_dir>/cmake-install/lib/legion_canonical_python.so"
```

Alternatively, if we installed into `${VIRTUAL_ENV}`, the generated `legion_canonical_cffi.py` would have this path:
```python
libpath = "${VIRTUAL_ENV}/lib/legion_canonical_python.so"
```

This leads to dlopen'ing a non-existent file at runtime, because the Legion Python bindings _do not_ install `liblegion_canonical_python.so` into `${VIRTUAL_ENV}/lib` -- for that we'd have to install all of Legion into `${VIRTUAL_ENV}`, defeating the purpose of editable installs.

### In a conda env

Install `_deps/legion-build/bindings/python` into `${CONDA_PREFIX}`, yielding the following CMake install rule:
```shell
${CONDA_PREFIX}/bin/python3 \
    -m pip install \
    --global-option=--cmake-build-dir "--global-option=<skbuild_dir>/cmake-build/_deps/legion-build/runtime" \
    --global-option=--prefix "--global-option=${CMAKE_INSTALL_PREFIX}" \
    --ignore-installed \
    --root / \
    --prefix ${CONDA_PREFIX} .
```

The above still installs the Legion Python bindings into `${CONDA_PREFIX}/<..>/site-packages`, and generates the following `legion_canonical_cffi.py`:
```python
libpath = "<skbuild_dir>/cmake-install/lib/legion_canonical_python.so"
```

This is the correct libpath when doing an `--editable` install in a conda env, but the exact opposite behavior from pip.

### Discrepencies

Let's revisit the Legion Python bindings' install rules for each flavor of venv:
```shell
# pip venv (not working):
python -m pip install ... --global-option=--prefix "--global-option=${CMAKE_INSTALL_PREFIX}" --prefix ${VIRTUAL_ENV} .
# pip venv (working):
python -m pip install ... --global-option=--prefix "--global-option=${CMAKE_INSTALL_PREFIX}" --prefix ${CMAKE_INSTALL_PREFIX} .
# conda env (working):
python -m pip install ... --global-option=--prefix "--global-option=${CMAKE_INSTALL_PREFIX}" --prefix ${CONDA_PREFIX} .
```

All of the above commands pass `--prefix` as a special "--global-option" thing, which (IIUC) is supposed to be forwarded on to `setup.py` and not handled by pip/setuptools. This should mean that the [parsed `--prefix` in Legion's setup.py](https://gitlab.com/StanfordLegion/legion/-/blob/master/bindings/python/setup.py?ref_type=heads#L14-15) should be the same. Except it's not.

* When installing into a pip venv, the `--prefix` `setup.py` sees is the bare `--prefix`
* When installing in a conda env, the `--prefix` `setup.py` sees is the wrapped `--global-option=--prefix`

:exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: :exploding_head: 

### In conclusion

This wouldn't be as painful if Legion's setup.py was using a proper PEP517 build backend, could supported `--editable` installs, etc. But as is, we're at the whims of pip and setuptools authors, and the builds are brittle and confusing.

File: install.py (hunk 1 above)
Comment (trxcllnt) [reply install.py:678]: @manopapad once we're using the devcontainers, I'd actually recommend using `all-major`. The object files are sensitive to the CUDA archs list, so if we're building with the same archs list, we'll all be populating the cache for each other/CI.

The devcontainers also come with some custom tuning to automatically parallelize device compilations (at the expense of total parallelism), so recompiling an individual object for all the archs completes in roughly the same time as building it for one arch.

**Comment by trxcllnt** [comment]
@Jacobfaib That's going to be a bit of a stretch, as this PR is rather urgent to unblock folks who need to use the latest `control_replication`. I plan on following up with additional workflow PRs once we finalize devcontainers in the next few weeks.

File: install.py (hunk 1 above)
Comment (trxcllnt) [reply install.py:678]: @manopapad are you sure the `--arch` flag doesn't work? This just compiled successfully for me:

```shell
~/legate/install.py --cuda --arch ampere --openmp --verbose --editable --clean
```

Comment (rohany) [reply install.py:678]: I pulled the most recent version of this branch and `--arch` doesn't work for me, the same error manolis is seeing

File: scripts/generate-conda-envs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (manopapad) [reply scripts/generate-conda-envs.py]: Not yet, so this discussion is kind of academic :-)

[the remark this answers is not in the document]
Comment (manopapad) [reply scripts/generate-conda-envs.py]: Let's just leave this as is, I was planning to do a pass over this file later anyway.

File: install.py (hunk 1 above)
Comment (manopapad) [reply install.py:678]: You are right, it works now.

**Review by manopapad — APPROVED** [review]
